crates/sp-domains/src/verification.rs

```diff
@@ -80,3 +96,156 @@ where
 
     Ok(())
 }
+
+pub fn verify_invalid_domain_extrinsics_root_fraud_proof<
+    CBlock,
+    DomainNumber,
+    DomainHash,
+    Balance,
+    Hashing,
+>(
+    consensus_state_root: CBlock::Hash,
+    bad_receipt: ExecutionReceipt<
+        NumberFor<CBlock>,
+        CBlock::Hash,
+        DomainNumber,
+        DomainHash,
+        Balance,
+    >,
+    fraud_proof: InvalidExtrinsicsRootProof,
+) -> Result<(), VerificationError>
+where
+    CBlock: Block,
+    CBlock::Hash: Into<H256>,
+    Hashing: Hasher<Out = CBlock::Hash>,
+{
+    let InvalidExtrinsicsRootProof {
+        valid_bundle_digests,
+        randomness_proof,
+        ..
+    } = fraud_proof;
+
+    let mut ext_values = Vec::new();
+    for (valid_bundle, bundle_digest) in bad_receipt
+        .valid_bundles
+        .into_iter()
+        .zip(valid_bundle_digests)
+    {
+        let bundle_digest_hash = BlakeTwo256::hash_of(&bundle_digest.bundle_digest);
+        if bundle_digest_hash != valid_bundle.bundle_digest {
+            return Err(VerificationError::InvalidBundleDigest);
+        }
+
+        ext_values.extend(bundle_digest.bundle_digest);
+    }
+
+    let storage_key = StorageKey(crate::fraud_proof::block_randomness_final_key());
+    let storage_proof = randomness_proof.clone();
+    let block_randomness = StorageProofVerifier::<Hashing>::verify_and_get_value::<Randomness>(
+        &consensus_state_root,
+        storage_proof,
+        storage_key,
+    )
+    .map_err(|_| VerificationError::InvalidProof)?;
+
+    let shuffling_seed = extrinsics_shuffling_seed::<Hashing>(block_randomness)
+        .into()
+        .to_fixed_bytes();
+    let ordered_extrinsics =
+        deduplicate_and_shuffle_extrinsics(ext_values, Randomness::from(shuffling_seed));
+    let ordered_trie_node_values = ordered_extrinsics
+        .iter()
+        .map(|ext_digest| match ext_digest {
+            ExtrinsicDigest::Data(data) => Value::Inline(data),
+            ExtrinsicDigest::Hash(hash) => Value::Node(hash.0.as_slice()),
+        })
+        .collect();
+    // TODO: include timestamp and domain runtime upgrade extrinsic
+    // TODO: change Layout version used for deriving extrinsics root in frame system
+    let extrinsics_root =
+        valued_ordered_trie_root::<LayoutV1<BlakeTwo256>>(ordered_trie_node_values);
+    if bad_receipt.domain_block_extrinsic_root == extrinsics_root {
+        return Err(VerificationError::InvalidProof);
+    }
+
+    Ok(())
+}
+
+fn extrinsics_shuffling_seed<Hashing>(block_randomness: Randomness) -> Hashing::Out
```

Comment:
If you need `Randomness` in the end, why is this function returning `Hashing::Out` if you have to convert it 3 times after that at all site?